File: vi/codes/csharp/chapter_backtracking/preorder_traversal_iii_template.cs
/**
 * File: preorder_traversal_iii_template.cs
 * Created Time: 2023-04-17
 * Author: hpstory (hpstory1024@163.com)
 */

namespace hello_algo.chapter_backtracking;

public class preorder_traversal_iii_template {
    /* Kiểm tra xem trạng thái hiện tại có phải là lời giải hay không */
    bool IsSolution(List<TreeNode> state) {
        return state.Count != 0 && state[^1].val == 7;
    }

    /* Ghi lại lời giải */
    void RecordSolution(List<TreeNode> state, List<List<TreeNode>> res) {
        res.Add(new List<TreeNode>(state));
    }

    /* Kiểm tra xem lựa chọn có hợp lệ trong trạng thái hiện tại hay không */
    bool IsValid(List<TreeNode> state, TreeNode choice) {
        return choice != null && choice.val != 3;
    }

    /* Cập nhật trạng thái */
    void MakeChoice(List<TreeNode> state, TreeNode choice) {
        state.Add(choice);
    }

    /* Khôi phục trạng thái */
    void UndoChoice(List<TreeNode> state, TreeNode choice) {
        state.RemoveAt(state.Count - 1);
    }

    /* Giải thuật quay lui: Ví dụ 3 */
    void Backtrack(List<TreeNode> state, List<TreeNode> choices, List<List<TreeNode>> res) {
        // Kiểm tra xem đây có phải là lời giải hay không
        if (IsSolution(state)) {
            // Ghi lại lời giải
            RecordSolution(state, res);
        }
        // Duyệt qua tất cả các lựa chọn
        foreach (TreeNode choice in choices) {
            // Cắt tỉa: kiểm tra xem lựa chọn có hợp lệ hay không
            if (IsValid(state, choice)) {
                // Thử: thực hiện lựa chọn, cập nhật trạng thái
                MakeChoice(state, choice);
                // Tiến hành vòng lựa chọn tiếp theo
                Backtrack(state, [choice.left!, choice.right!], res);
                // Quay lui: hoàn tác lựa chọn, khôi phục về trạng thái trước đó
                UndoChoice(state, choice);
            }
        }
    }

    [Test]
    public void Test() {
        TreeNode? root = TreeNode.ListToTree([1, 7, 3, 4, 5, 6, 7]);
        Console.WriteLine("\nInitialize binary tree");
        PrintUtil.PrintTree(root);

        // Giải thuật quay lui
        List<List<TreeNode>> res = [];
        List<TreeNode> choices = [root!];
        Backtrack([], choices, res);

        Console.WriteLine("\nOutput all paths from root node to node 7, requiring paths do not include nodes with value 3");
        foreach (List<TreeNode> path in res) {
            PrintUtil.PrintList(path.Select(p => p.val).ToList());
        }
    }
}

File: vi/codes/go/chapter_backtracking/preorder_traversal_iii_template.go
// File: preorder_traversal_iii_template.go
// Created Time: 2023-05-09
// Author: Reanon (793584285@qq.com)

package chapter_backtracking

import (
	. "github.com/krahets/hello-algo/pkg"
)

/* Kiểm tra xem state hiện tại có phải là lời giải hay không */
func isSolution(state *[]*TreeNode) bool {
	return len(*state) != 0 && (*state)[len(*state)-1].Val == 7
}

/* Ghi lại lời giải */
func recordSolution(state *[]*TreeNode, res *[][]*TreeNode) {
	*res = append(*res, append([]*TreeNode{}, *state...))
}

/* Kiểm tra xem lựa chọn có hợp lệ dưới state hiện tại hay không */
func isValid(state *[]*TreeNode, choice *TreeNode) bool {
	return choice != nil && choice.Val != 3
}

/* Cập nhật state */
func makeChoice(state *[]*TreeNode, choice *TreeNode) {
	*state = append(*state, choice)
}

/* Khôi phục state */
func undoChoice(state *[]*TreeNode, choice *TreeNode) {
	*state = (*state)[:len(*state)-1]
}

/* Giải thuật quay lui: ví dụ 3 */
func backtrackIII(state *[]*TreeNode, choices *[]*TreeNode, res *[][]*TreeNode) {
	// Kiểm tra xem đây có phải là lời giải hay không
	if isSolution(state) {
		// Ghi lại lời giải
		recordSolution(state, res)
	}
	// Duyệt qua tất cả các lựa chọn
	for _, choice := range *choices {
		// Cắt tỉa: kiểm tra xem lựa chọn có hợp lệ hay không
		if isValid(state, choice) {
			// Thử: thực hiện lựa chọn, cập nhật state
			makeChoice(state, choice)
			// Tiến hành vòng lựa chọn tiếp theo
			temp := make([]*TreeNode, 0)
			temp = append(temp, choice.Left, choice.Right)
			backtrackIII(state, &temp, res)
			// Quay lui: hoàn tác lựa chọn, khôi phục về state trước đó
			undoChoice(state, choice)
		}
	}
}
